fabric-tool-attribute-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/AttributeManager.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.tool.attribute;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+import net.minecraft.entity.attribute.EntityAttributeModifier;
+
+public final class AttributeManager {
+	private AttributeManager() { }
+
+	/**
+	 * Merge two multimaps of EntityAttributeModifiers, combining their modifiers to prevent duplicate entries in a tooltip.
+	 *
+	 * @param left  The first of the two multimaps to merge.
+	 * @param right The second of the two multimaps to merge.
+	 * @return The merged form of the two.
+	 */
+	public static Multimap<String, EntityAttributeModifier> mergeAttributes(Multimap<String, EntityAttributeModifier> left, Multimap<String, EntityAttributeModifier> right) {
+		if (left.isEmpty()) return right;
+		if (right.isEmpty()) return left;
+		Multimap<String, EntityAttributeModifier> ret = HashMultimap.create();
+		Set<String> allKeys = new HashSet<>();
+		allKeys.addAll(left.keySet());
+		allKeys.addAll(right.keySet());
+
+		for (String key : allKeys) {
+			double add = 0;
+			double multBase = 0;
+			double multTotal = 0;
+			Collection<EntityAttributeModifier> modifiers;
+
+			if (left.containsKey(key)) {
+				modifiers = left.get(key);
+				if (right.containsKey(key)) modifiers.addAll(right.get(key));
+			} else {
+				//key *must* be in either left or right, so if it's not in left, it *must* be in right
+				modifiers = right.get(key);
+			}
+
+			for (EntityAttributeModifier mod : modifiers) {
+				switch (mod.getOperation()) {
+				case ADDITION:
+					add += mod.getAmount();
+					break;
+				case MULTIPLY_BASE:
+					multBase += mod.getAmount();
+					break;
+				case MULTIPLY_TOTAL:
+					multTotal += mod.getAmount();
+					break;
```

Comment:
Are you sure we want to combine `MULTIPLY_BASE` and `MULTIPLY_TOTAL` like this? It seems rather odd, especially since vanilla does it differently iirc.

Reply:
I'll double-check, but I'm pretty sure that this is how vanilla does it.

Reply:
I'm gonna need to get a pen and paper to figure this out. It looks *pretty* similar to what vanilla does, but I wanna verify.

Reply:
god, EAMs are so weird and hacky. There's two local vars in totaling, `d` and `e`.  It goes through these steps:
1. `d` starts at the base value of the attribute, then iterates over all the `ADDITION` modifiers, adding them each to `d`.
2. `e` is set to the value of `d`, then iterates over all the `MULTIPLY_BASE` modifiers, with the operation `e += d * modifier.getAmount()` .
3. it then iterates over all the `MUTLIPLY_TOTAL` modifiers, with the operation `e *= 1.0D + entityAttributeModifier3.getAmount()`.
4. lastly, it clamps the value based on the attribute's min and max.

I'm thinking we *might* be able to use the associative property on `MULTIPLY_BASE` but I just really can't tell for sure unless I do some pretty rigorous testing.

Reply:
alright, testing proves this doesn't work. I'm not sure how to actually *make* it work, is the fundamental issue.

Reply:
I've gotten `MULTIPLY_TOTAL` to work properly, but I'm not sure it's even *possible* to get `MULTIPLY_BASE` to work, due to the operations it performs. My two options are to run low or run high. Which do you think is better?

Reply:
holy shit, I found a way to make the math work perfectly!